redfish-core/lib/log_services.hpp

```diff
@@ -6606,4 +6613,581 @@ inline void requestRoutesSystemHardwareIsolationLogService(App& app)
+    if (messageArgs.size() > 0)
+    {
+        int i = 0;
+        for (auto messageArg : messageArgs)
+        {
+            std::string argStr = "%" + std::to_string(++i);
+            size_t argPos = msg.find(argStr);
+            if (argPos != std::string::npos)
+            {
+                msg.replace(argPos, argStr.length(), messageArg);
+            }
+        }
+    }
+
+    // Fill in the log entry with the gathered data
+    logEntryJson["@odata.type"] = "#LogEntry.v1_9_0.LogEntry";
+    logEntryJson["@odata.id"] =
+        "/redfish/v1/Systems/system/LogServices/AuditLog/Entries/" + logEntryID;
+    logEntryJson["Name"] = "Audit Log Entry";
+    logEntryJson["Id"] = logEntryID;
+    logEntryJson["MessageId"] = std::move(messageID);
+    logEntryJson["Message"] = std::move(msg);
+    logEntryJson["MessageArgs"] = std::move(messageArgs);
+    logEntryJson["EntryType"] = "Event";
+    logEntryJson["EventTimestamp"] = std::move(entryTimeStr);
+
+    /* logEntryJson["Oem"]["IBM"]["@odata.id"] ?? */
+    logEntryJson["Oem"]["IBM"]["@odata.type"] = "#OemLogEntry.v1_0_0.LogEntry";
+    logEntryJson["Oem"]["IBM"]["AdditionalDataFullAuditLogURI"] =
+        "/redfish/v1/Systems/system/LogServices/AuditLog/Entries/FullAudit/attachment";
+
+    return LogParseError::success;
+}
+
+/**
+ * @brief Read single line from POSIX stream
+ *
+ * @param[in] logStream - POSIX stream
+ * @param[out] logLine - Filled in with line read on success
+ *
+ * @return True if line was read. False if EOF reached or line
+ *         exceeds 4096 characters
+ */
+inline bool readLine(FILE* logStream, std::string& logLine)
+{
+    std::array<char, 4096> buffer;
+
+    if (fgets(buffer.data(), buffer.size(), logStream) == nullptr)
+    {
+        /* Failed to read, could be EOF */
+        return false;
+    }
+
+    logLine.assign(buffer.data());
+
+#if 0
+    /* Could repeat reads or increase size of buffer.
+     * Don't expect log lines to be longer than 4096 so
+     * erroring out to protect against malformed data.
+     */
+    if (!logLine.ends_with('\n'))
+    {
+            /* Line was too long for the buffer.
+             * Try one more read for rest of the line.
+             */
+        if (fgets(buffer.data(), buffer.size(), logStream) != nullptr)
+        {
+            logLine.append(buffer.data());
+        }
+    }
+#endif
+
+    /* Return success only if we got a complete line */
+    return logLine.ends_with('\n');
+}
+
+/**
+ * @brief Reads the audit log entries and writes them to Members array
+ *
+ * @param[in] asyncResp - The redfish response to return.
+ * @param[in] unixfd - File descriptor for Audit Log file
+ * @param[in] skip - Query paramater skip value
+ * @param[in] top - Query paramater top value
+ */
+void readAuditLogEntries(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                         const sdbusplus::message::unix_fd& unixfd, size_t skip,
+                         size_t top)
+{
+    auto fd = dup(unixfd);
```

Comment:
I took a look at the upstream use and the code for the boost::beast::file_posix class. I agree using a class here would make cleanup safer.

However, the file_posix class does not have a constructor or method which will dup() a passed in unixfd. It will just inherit the one passed to it. So making this change would change the semantics. The dup() is being done since the D-Bus is the owner of the returned unix_fd and need to make sure it isn't closed underneath us while we are still processing it. Also want to own the SEEK position of the fd.

There is also still upstream code which does the dup() of the unix_fd. (https://github.com/openbmc/bmcweb/blob/2ae81db99a155febe45d27e3834cfa5471b3ecdf/redfish-core/lib/log_services.hpp#L806)

I'll put this on my list of upstream things to consider changing. But for now I will keep this as-is.

Reply:
I took a look at the boost::beast::file_posix class. I agree that it would help with cleanup if I was just using the fd as-is. However I am associating a stream in order to simplify the readLine() function. The file_posix class does not have a stream option.

The closing of a stream ([fclose()](https://man7.org/linux/man-pages/man3/fclose.3p.html)) also closes its related file descriptor. This would mean the fd associated with the posix_file class would get closed outside of the destructor or close call for the created entity.

I will leave this as-is for now. I will add to my upstream work a todo item to consider making a class for the stream I'm using to make the cleanup neater. I think I would make the dup() and fdopen() all part of the constructor.

Reply:
We can keep using `fd` as is after dup().
Or you can get `fd` using `fileHandle.native_handle()`.

The key reason to suggest this is to prevent the mistake by not doing `close()` or by the exception throw.

By doing this code block, the existing code can remain same but `close(fd)` is no longer needed.

```
#include <boost/beast/core/file_posix.hpp>
boost::beast::file_posix fileHandle;
int fd = dup(unixfd);
fileHandle.native_handle( fd );

```

Reply:
I am not seeing how the use of that class here would be better and easier to understand.

Looking at the destructor code for file_posix() it will close the fd associated with it. (https://github.com/boostorg/beast/blob/e0a9da265f63daf70d9db61eb20c6b4141c6fead/include/boost/beast/core/impl/file_posix.ipp#L74)

But the fclose() for the logStream is also going to close the same fd. The file_posix class copy of that fd is not going to have been cleared and is not going to understand that it was already closed.

To use this class with a stream would require extra hoops that I think would make the code less easier to understand.

Reply:
If we use the class `file_posix`, like any C++ class handling,  `fd` in `fileHandle` will be closed automatically by  its destructor (i.e. exical scope exit by function return, or exception throw). 
If there are places which need to explicitly close it (e.g. fclose()),  `fileHandle.native_handle(-1)` can be set to free it.

Either way is good for now.  I would think that this may need to be changed while doing upstreaming....